pkg/services/ngalert/state/manager.go

```diff
@@ -279,11 +283,13 @@ func (st *Manager) setNextState(ctx context.Context, alertRule *ngModels.AlertRu
 
 	st.set(currentState)
 
+	var old *InstanceStateAndReason
 	shouldUpdateAnnotation := oldState != currentState.State || oldReason != currentState.StateReason
 	if shouldUpdateAnnotation {
+		old = &InstanceStateAndReason{State: oldState, Reason: oldReason}
 		go st.annotateState(ctx, alertRule, currentState.Labels, result.EvaluatedAt, InstanceStateAndReason{State: currentState.State, Reason: currentState.StateReason}, InstanceStateAndReason{State: oldState, Reason: oldReason})
 	}
-	return currentState
+	return currentState, old
```

Comment:
Nit: Can we always return an `InstanceStateAndReason{State: oldState, Reason: oldReason}` from `setNextState`, and then above in `ProcessEvalResults` have:
```
if previous.State != currentState.State || previous.Reason != currentState.StateReason || s.State == eval.Pending {
    statesToSave = append(statesToSave, s)
}
```
All this really does is lift the if condition, but it's more clear as to what's going on IMO. You can see at a glance exactly under what conditions we write to DB, without having to deeply understand code in two separate spots.